include/dsn/dist/replication/replication_app_client_base.h

```diff
@@ -323,19 +334,56 @@ namespace dsn { namespace replication {
                 owner
                 );
 
-            auto rc = create_read_context(partition_index, code, msg, task, read_semantic, snapshot_decree, reply_hash);
+            auto rc = create_read_context(-1, key_hash, code, msg, task, read_semantic, snapshot_decree, reply_hash);
             ::marshall(msg, req);
             call(rc);
             return std::move(task);
         }
 
-        // get read address policy
-        virtual ::dsn::rpc_address get_read_address(read_semantic_t semantic, const partition_configuration& config);
+        template<typename TRequest, typename TResponse>
+        ::dsn::task_ptr request_meta(
+            dsn_task_code_t code,
+            std::shared_ptr<TRequest>& req,
+
+            // callback
+            clientlet* owner,
+            std::function<void(error_code, std::shared_ptr<TRequest>&, std::shared_ptr<TResponse>&)> callback,
+
+            // other specific parameters
+            int timeout_milliseconds= 0,
+            int reply_hash = 0
+            )
+        {
+            dsn_message_t msg = dsn_msg_create_request(code, timeout_milliseconds, 0);
+            ::marshall(msg, *req);
+
+            task_ptr task = ::dsn::rpc::internal_use_only::create_rpc_call(
+                msg,
+                req,
+                callback,
+                reply_hash,
+                owner
+                );
+
+            meta_context* context = new meta_context;
+            context->callback_task = task;
+            context->completed = false;
+            context->request = msg;
+            context->timeout_timer = nullptr;
+            context->timeout_ts_us = now_us() + timeout_milliseconds * 1000;
+
+            call_meta(context);
+            return std::move(task);
+        }
+
         ::dsn::rpc_address get_meta_servers() const { return _meta_servers; }
+
     public:
+
         struct request_context : public ref_counter
         {
             int                   partition_index;
```

Comment:
is there any place that the upper applications know the partition id explicitly? I think we have agreed that the upper apps only define hash value, and the underlying lib will translate that into partition id implicitly.